test/publish.test.js

```diff
@@ -176,6 +176,133 @@ test.serial('Publish a release with an asset with a template label', async (t) =
+  const options = {repositoryUrl: `https://gitlab.com/${owner}/${repo}.git`};
+  const encodedRepoId = encodeURIComponent(`${owner}/${repo}`);
+  const encodedGitTag = encodeURIComponent(nextRelease.gitTag);
+  const assets = [
+    {
+      label: 'README.md',
+      type: `\${process.env.TYPE}`,
+      url: 'https://gitlab.com/gitlab-org/gitlab/-/blob/master/README.md',
+    },
+  ];
+  const gitlab = authenticate(env)
+    .post(`/projects/${encodedRepoId}/releases`, {
+      tag_name: nextRelease.gitTag,
+      description: nextRelease.notes,
+      assets: {
+        links: [
+          {
+            name: 'README.md',
+            url: `https://gitlab.com/gitlab-org/gitlab/-/blob/master/README.md`,
+            link_type: 'other',
+          },
+        ],
+      },
+    })
+    .reply(200);
+
+  const result = await publish({assets}, {env, cwd, options, nextRelease, logger: t.context.logger});
+
+  t.is(result.url, `https://gitlab.com/${owner}/${repo}/-/releases/${encodedGitTag}`);
+  t.deepEqual(t.context.log.args[0], ['Published GitLab release: %s', nextRelease.gitTag]);
+  t.true(gitlab.isDone());
+});
+
+test.serial('Publish a release with an asset with a template filepath', async (t) => {
+  process.env.FILEPATH = '/dist/file.css';
+  const cwd = 'test/fixtures/files';
+  const owner = 'test_user';
+  const repo = 'test_repo';
+  const env = {GITLAB_TOKEN: 'gitlab_token'};
+  const nextRelease = {gitHead: '123', gitTag: 'v1.0.0', notes: 'Test release note body', version: '1.0.0'};
+  const options = {repositoryUrl: `https://gitlab.com/${owner}/${repo}.git`};
+  const encodedRepoId = encodeURIComponent(`${owner}/${repo}`);
+  const encodedGitTag = encodeURIComponent(nextRelease.gitTag);
+  const uploaded = {url: '/uploads/file.css', alt: 'file.css'};
+  const assets = [
+    {
+      label: `file.css`,
+      path: 'file.css',
+      type: 'other',
+      filepath: `\${process.env.FILEPATH}`,
+    },
+  ];
+  const gitlab = authenticate(env)
+    .post(`/projects/${encodedRepoId}/releases`, {
+      tag_name: nextRelease.gitTag,
+      description: nextRelease.notes,
+      assets: {
+        links: [
+          {
+            name: 'file.css',
+            url: `https://gitlab.com/${owner}/${repo}${uploaded.url}`,
+            link_type: 'other',
+            filepath: '/dist/file.css',
+          },
+        ],
+      },
+    })
+    .reply(200);
+  const gitlabUpload = authenticate(env)
+    .post(`/projects/${encodedRepoId}/uploads`, /filename="file.css"/gm)
+    .reply(200, uploaded);
+
+  const result = await publish({assets}, {env, cwd, options, nextRelease, logger: t.context.logger});
+
+  t.is(result.url, `https://gitlab.com/${owner}/${repo}/-/releases/${encodedGitTag}`);
+  t.deepEqual(t.context.log.args[0], ['Uploaded file: %s', uploaded.url]);
+  t.deepEqual(t.context.log.args[1], ['Published GitLab release: %s', nextRelease.gitTag]);
+  t.true(gitlabUpload.isDone());
+  t.true(gitlab.isDone());
+});
+
```

Comment:
These tests are super-similar, what about just having one that we parameterize?

Reply:
Do you have any example how to make a parameterized test in JS?
I only have experience with Java for this scenario.

Reply:
Ava, our test runner, [does not have a built-in way to do this](https://github.com/avajs/ava/issues/2414). Either wrap the entire `test.serial(...)` in a loop or add [`pava`](https://github.com/TomerAberbach/pava)

Reply:
What do you think about putting all in one test? (: -- I mean its nearly the same and for me more like an integration rather than an unit test.

```
test.serial('Publish a release (with an link) with variables', async (t) => {
  process.env.TYPE = 'other';
  const cwd = 'test/fixtures/files';
  const owner = 'test_user';
  const repo = 'test_repo';
  const env = {GITLAB_TOKEN: 'gitlab_token'};
  const nextRelease = {gitHead: '123', gitTag: 'v1.0.0', notes: 'Test release note body', version: '1.0.0'};
  const options = {repositoryUrl: `https://gitlab.com/${owner}/${repo}.git`};
  const encodedRepoId = encodeURIComponent(`${owner}/${repo}`);
  const encodedGitTag = encodeURIComponent(nextRelease.gitTag);
  const assets = [
    {
      label: 'README-v\${nextRelease.version}.md',
      type: `\${process.env.TYPE}`,
      url: `https://gitlab.com/gitlab-org/gitlab/-/blob/master/README-v\${nextRelease.version}.md`,
    },
  ];
  const gitlab = authenticate(env)
    .post(`/projects/${encodedRepoId}/releases`, {
      tag_name: nextRelease.gitTag,
      description: nextRelease.notes,
      assets: {
        links: [
          {
            name: 'README-v1.0.0.md',
            url: `https://gitlab.com/gitlab-org/gitlab/-/blob/master/README-v1.0.0.md`,
            link_type: 'other',
          },
        ],
      },
    })
    .reply(200);

  const result = await publish({assets}, {env, cwd, options, nextRelease, logger: t.context.logger});

  t.is(result.url, `https://gitlab.com/${owner}/${repo}/-/releases/${encodedGitTag}`);
  t.deepEqual(t.context.log.args[0], ['Published GitLab release: %s', nextRelease.gitTag]);
  t.true(gitlab.isDone());
});
```